src/main/java/org/folio/circulation/services/EventPublishingService.java

```diff
@@ -0,0 +1,164 @@
+package org.folio.circulation.services;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.support.JsonPropertyWriter.write;
+import static org.folio.circulation.support.Result.succeeded;
+import static org.folio.rest.util.OkapiConnectionParams.OKAPI_TENANT_HEADER;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.folio.circulation.domain.CheckInProcessRecords;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.RequestAndRelatedRecords;
+import org.folio.circulation.resources.TenantAPI;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.folio.rest.jaxrs.model.Event;
+import org.folio.rest.jaxrs.model.EventMetadata;
+import org.folio.rest.util.OkapiConnectionParams;
+import org.folio.util.pubsub.PubSubClientUtils;
+
+import io.vertx.core.Context;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.RoutingContext;
+
+public class EventPublishingService {
+  private static final Logger logger = LoggerFactory.getLogger(TenantAPI.class);
+
+  public static final String ITEM_CHECKED_OUT_EVENT_TYPE = "ITEM_CHECKED_OUT";
+  public static final String ITEM_CHECKED_IN_EVENT_TYPE = "ITEM_CHECKED_IN";
+  public static final String ITEM_DECLARED_LOST_EVENT_TYPE = "ITEM_DECLARED_LOST";
+  public static final String LOAN_DUE_DATE_UPDATED_EVENT_TYPE = "LOAN_DUE_DATE_UPDATED";
+
+  private final Map<String, String> okapiHeaders;
+  private final Context vertxContext;
+
+  private Clients clients;
+
+  public EventPublishingService(RoutingContext routingContext) {
+    okapiHeaders = routingContext.request().headers().entries().stream()
+      .collect(Collectors.toMap(entry -> entry.getKey().toLowerCase(), Map.Entry::getValue));
+    vertxContext = routingContext.vertx().getOrCreateContext();
+  }
+
+  public EventPublishingService(RoutingContext routingContext, Clients circulationClients) {
+    this(routingContext);
+    clients = circulationClients;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> publishItemCheckedOutEvent(
+    LoanAndRelatedRecords loanAndRelatedRecords) {
+
+    if (loanAndRelatedRecords.getLoan() != null) {
+      Loan loan = loanAndRelatedRecords.getLoan();
+
+      JsonObject payloadJsonObject = new JsonObject();
```

Comment:
The event name describes an event of check out an item. It seems to me that payload should contain as less custom data as possible (in this case only `item id`) and subscribers have to fetch all needed additional data themselves.  

Otherwise, we're introducing dependencies on the payload version between all subscribers.

Reply:
I believe one of the reasons for using messaging was to avoid dependencies (including circular dependencies) between modules. If a subscriber needs to call back for more details it would undermine at least part of the idea.
But I agree that we're introducing another dependency. I'm not aware of any plans for using Schema Registry or something similar for message versioning.

Reply:
Good point, I did not think about circular dependencies. 

Resolving comment.